src/DependencyInjection/DI/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -358,6 +359,103 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
             return parameterCallSites;
         }
 
+        private static bool IsCompatibleWithGenericParameterConstraints(Type genericTypeDefinition, Type[] parameters)
+        {
+            var genericArgumentDefinitions = genericTypeDefinition.GetTypeInfo().GenericTypeParameters;
+            for (var i = 0; i < genericArgumentDefinitions.Length; ++i)
+            {
+                var argumentDefinitionTypeInfo = genericArgumentDefinitions[i].GetTypeInfo();
+                var parameter = parameters[i];
+                var parameterTypeInfo = parameter.GetTypeInfo();
+                if (argumentDefinitionTypeInfo.GetGenericParameterConstraints()
+                   .Select(constraint => SubstituteGenericParameterConstraint(parameters, constraint))
+                   .Any(constraint => !ParameterCompatibleWithTypeConstraint(parameter, constraint)))
+                {
+                    return false;
+                }
+                var specialConstraints = argumentDefinitionTypeInfo.GenericParameterAttributes;
+                if ((specialConstraints & GenericParameterAttributes.DefaultConstructorConstraint)
+                   == GenericParameterAttributes.DefaultConstructorConstraint)
```

Comment:
Join lines.

Reply:
Fixed! And others with multi-line conditionals.